File: bindings/node/examples/basic.ts
/**
 * Ainos SDK — Basic usage examples.
 *
 * Run with: `npx ts-node examples/basic.ts`
 *
 * These examples demonstrate:
 * - Basic inference
 * - Streaming inference
 * - Model management
 * - Context store
 * - Error handling
 * - Authentication
 * - Batch inference
 * - Health checks
 */

import {
  AinosClient,
  createClient,
  ConnectionError,
  AuthError,
  InferenceError,
  TimeoutError,
  RateLimitError,
  DaemonError,
  accumulateStream,
} from '../src';

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

// These can be overridden with environment variables
const HOST = process.env.AINOS_HOST || '127.0.0.1';
const PORT = parseInt(process.env.AINOS_PORT || '9500', 10);
const AUTH_TOKEN = process.env.AINOS_TOKEN || '';

// ============================================================================
// Example 1: Basic Inference
// ============================================================================

async function exampleBasicInference(): Promise<void> {
  console.log('\n=== Example 1: Basic Inference ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
    connectTimeout: 5000,
    readTimeout: 30000,
  });

  try {
    await client.connect();
    console.log('Connected to Ainos daemon');

    const response = await client.infer({
      prompt: 'What is Ainos OS?',
      model: 'default',
      temperature: 0.7,
      maxTokens: 256,
    });

    console.log('Response:', response.output);
    console.log(`Tokens: ${response.tokensGenerated}, Time: ${response.inferenceMs}ms, Source: ${response.source}`);
  } catch (err) {
    console.error('Inference error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 2: Streaming Inference
// ============================================================================

async function exampleStreamingInference(): Promise<void> {
  console.log('\n=== Example 2: Streaming Inference ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();
    console.log('Connected, starting streaming inference...\n');

    const stream = client.inferStream({
      prompt: 'Write a short poem about artificial intelligence.',
      model: 'default',
      temperature: 0.8,
      maxTokens: 500,
    });

    // Accumulate for display
    let fullText = '';

    stream.on('data', (chunk: string) => {
      process.stdout.write(chunk);
      fullText += chunk;
    });

    stream.on('progress', (tokens: number, elapsed: number) => {
      // Progress updates
    });

    stream.on('end', () => {
      console.log('\n\n--- Stream complete ---');
      console.log(`Total length: ${fullText.length} chars`);
    });

    stream.on('error', (err: Error) => {
      console.error('\nStream error:', err);
    });

    // Wait for stream to complete
    await stream.waitForEnd();
  } catch (err) {
    console.error('Streaming error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 3: Accumulate Stream to String
// ============================================================================

async function exampleAccumulateStream(): Promise<void> {
  console.log('\n=== Example 3: Accumulate Stream to String ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    const stream = client.inferStream({
      prompt: 'Explain machine learning in one paragraph.',
      maxTokens: 200,
    });

    const fullText = await accumulateStream(stream);
    console.log('Full output:', fullText);
    console.log(`Length: ${fullText.length} chars`);
  } catch (err) {
    console.error('Error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 4: Model Management
// ============================================================================

async function exampleModelManagement(): Promise<void> {
  console.log('\n=== Example 4: Model Management ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    // List models
    const models = await client.modelList();
    console.log(`Available models (${models.length}):`);
    for (const model of models) {
      const status = model.loaded ? 'loaded' : 'unloaded';
      console.log(`  - ${model.id} (${model.name}) [${status}] ${model.sizeMb}MB`);
    }

    // Load a model (if a path is provided)
    const modelPath = process.env.AINOS_MODEL_PATH;
    if (modelPath) {
      console.log(`\nLoading model from: ${modelPath}`);
      const loadResult = await client.modelLoad(modelPath);
      console.log(`Load result: ${loadResult.status} — ${loadResult.message}`);

      if (loadResult.modelInfo) {
        console.log(`Model ID: ${loadResult.modelInfo.id}`);
        console.log(`Architecture: ${loadResult.modelInfo.architecture}`);
      }

      // Unload the model
      if (loadResult.modelId) {
        console.log(`\nUnloading model: ${loadResult.modelId}`);
        await client.modelUnload(loadResult.modelId);
        console.log('Model unloaded successfully');
      }
    }
  } catch (err) {
    console.error('Model management error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 5: System Status
// ============================================================================

async function exampleSystemStatus(): Promise<void> {
  console.log('\n=== Example 5: System Status ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    const status = await client.status();
    console.log('Daemon Status:');
    console.log(`  Uptime: ${status.uptime}s`);
    console.log(`  Models loaded: ${status.modelsLoaded}`);
    console.log(`  Total requests: ${status.totalRequests}`);
    console.log(`  Network available: ${status.networkAvailable}`);
    if (status.activeSessions !== undefined) {
      console.log(`  Active sessions: ${status.activeSessions}`);
    }
    if (status.rateLimits) {
      console.log('  Rate limits:');
      for (const rl of status.rateLimits) {
        console.log(`    ${rl.category}: ${rl.remaining}/${rl.limit} (resets in ${rl.resetSeconds}s)`);
      }
    }

    // Health check
    const health = await client.health();
    console.log(`\nHealth: ${health.ok ? 'OK' : 'FAIL'}`);
    if (health.uptime !== undefined) {
      console.log(`Daemon uptime: ${health.uptime}s`);
    }
  } catch (err) {
    console.error('Status error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 6: Context Store
// ============================================================================

async function exampleContextStore(): Promise<void> {
  console.log('\n=== Example 6: Context Store ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    const sessionId = 'example-session';
    const key = 'user-preferences';
    const value = JSON.stringify({ theme: 'dark', language: 'en' });

    // Store
    await client.contextStore(sessionId, key, value);
    console.log(`Stored context: ${key}`);

    // Retrieve
    const retrieved = await client.contextRetrieve(sessionId, key);
    if (retrieved) {
      console.log(`Retrieved: ${retrieved.toString('utf-8')}`);
    } else {
      console.log('Key not found');
    }

    // Retrieve non-existent
    const missing = await client.contextRetrieve(sessionId, 'nonexistent-key');
    console.log(`Non-existent key: ${missing === null ? 'null (correct)' : 'unexpected'}`);
  } catch (err) {
    console.error('Context store error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 7: Batch Inference
// ============================================================================

async function exampleBatchInference(): Promise<void> {
  console.log('\n=== Example 7: Batch Inference ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    const prompts = [
      'What is the capital of France?',
      'What is 2 + 2?',
      'Who wrote Romeo and Juliet?',
    ];

    const requests = prompts.map((prompt) => ({
      prompt,
      maxTokens: 50,
    }));

    const responses = await client.batchInfer(requests);
    console.log(`Processed ${responses.length} requests:`);
    for (let i = 0; i < responses.length; i++) {
      console.log(`  [${i}] ${prompts[i]}`);
      console.log(`      -> ${responses[i].output.slice(0, 100)}...`);
    }
  } catch (err) {
    console.error('Batch inference error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 8: Rate Limit Status
// ============================================================================

async function exampleRateLimitStatus(): Promise<void> {
  console.log('\n=== Example 8: Rate Limit Status ===');

  const client = new AinosClient({
    host: HOST,
    port: PORT,
    authToken: AUTH_TOKEN,
  });

  try {
    await client.connect();

    const rateLimit = await client.rateLimitStatus();
    console.log('Rate limit status:');
    for (const limit of rateLimit.limits) {
      console.log(`  ${limit.category}: ${limit.remaining}/${limit.limit} remaining (resets in ${limit.resetSeconds}s)`);
    }
  } catch (err) {
    console.error('Rate limit error:', err);
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 9: Error Handling
// ============================================================================

async function exampleErrorHandling(): Promise<void> {
  console.log('\n=== Example 9: Error Handling ===');

  // Try connecting to a non-existent daemon
  const client = new AinosClient({
    host: '127.0.0.1',
    port: 19999, // Wrong port
    connectTimeout: 2000,
    autoReconnect: false,
  });

  try {
    await client.connect();
    console.log('Connected (unexpected)');
  } catch (err) {
    if (err instanceof ConnectionError) {
      console.log('Caught ConnectionError:', err.message);
    } else if (err instanceof TimeoutError) {
      console.log('Caught TimeoutError:', err.message);
    } else {
      console.log('Caught error:', err);
    }
  } finally {
    client.disconnect();
  }
}

// ============================================================================
// Example 10: Factory Function
// ============================================================================

async function exampleFactoryFunction(): Promise<void> {
  console.log('\n=== Example 10: Factory Function ===');

  try {
    const client = await createClient({
      host: HOST,
      port: PORT,
      authToken: AUTH_TOKEN,
      connectTimeout: 5000,
    });

    console.log('Client created and connected via factory');

    const response = await client.infer({
      prompt: 'Hello!',
      maxTokens: 100,
    });

    console.log('Response:', response.output);
    client.disconnect();
  } catch (err) {
    console.error('Factory error:', err);
  }
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  console.log('Ainos SDK Examples');
  console.log(`Target: ${HOST}:${PORT}`);
  console.log(`Auth: ${AUTH_TOKEN ? 'configured' : 'not configured'}`);
  console.log('='.repeat(50));

  const examples = [
    exampleBasicInference,
    exampleStreamingInference,
    exampleAccumulateStream,
    exampleModelManagement,
    exampleSystemStatus,
    exampleContextStore,
    exampleBatchInference,
    exampleRateLimitStatus,
    exampleErrorHandling,
    exampleFactoryFunction,
  ];

  // Determine which examples to run
  const specificExample = process.argv[2];
  if (specificExample) {
    const idx = parseInt(specificExample, 10);
    if (idx >= 1 && idx <= examples.length) {
      await examples[idx - 1]();
    } else {
      console.error(`Example ${idx} not found. Choose 1-${examples.length}.`);
    }
  } else {
    // Run all examples (skip error handling to avoid confusion)
    for (let i = 0; i < examples.length; i++) {
      try {
        await examples[i]();
      } catch (err) {
        console.error(`Example ${i + 1} failed:`, err);
      }
    }
  }

  console.log('\nDone.');
}

main().catch(console.error);

File: bindings/rust/ainos-sdk/examples/basic.rs
//! Basic usage examples for the Ainos SDK.
//!
//! This file demonstrates all major features of the SDK, including:
//! - Basic inference
//! - Streaming inference
//! - Model management (list, load, unload)
//! - Context operations
//! - Status queries
//! - Error handling
//! - Connection management
//!
//! Run with: `cargo run --example basic`

use ainos_sdk::error::{Result, Retryable};
use ainos_sdk::prelude::*;
use std::time::Duration;

// ---------------------------------------------------------------------------
// Helper: create a client
// ---------------------------------------------------------------------------

/// Create a client connected to the daemon.
///
/// Reads the `AINOS_HOST`, `AINOS_PORT`, and `AINOS_TOKEN` environment
/// variables, falling back to sensible defaults.
async fn create_client() -> Result<AinosClient> {
    let host = std::env::var("AINOS_HOST").unwrap_or_else(|_| "127.0.0.1".to_string());
    let port: u16 = std::env::var("AINOS_PORT")
        .unwrap_or_else(|_| "9500".to_string())
        .parse()
        .expect("AINOS_PORT must be a number");
    let token = std::env::var("AINOS_TOKEN").ok();

    let mut builder = AinosClient::builder()
        .host(host)
        .port(port)
        .connect_timeout(Duration::from_secs(10))
        .read_timeout(Duration::from_secs(120));

    if let Some(t) = token {
        builder = builder.auth_token(t);
    }

    let client = builder.build();
    client.connect().await?;
    Ok(client)
}

// ---------------------------------------------------------------------------
// Example 1: Basic inference
// ---------------------------------------------------------------------------

/// Send a simple inference request and print the response.
async fn example_basic_inference(client: &AinosClient) -> Result<()> {
    println!("\n=== Basic Inference ===");

    let req = InferenceRequest::builder()
        .prompt("What is the meaning of life?")
        .model("default")
        .temperature(0.7)
        .max_tokens(200)
        .build();

    let resp = client.infer(&req).await?;

    println!("Prompt: {}", req.prompt);
    println!("Response: {}", resp.output);
    println!("Tokens generated: {}", resp.tokens_generated);
    println!("Inference time: {} ms", resp.inference_ms);
    println!("Source: {}", resp.source);

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 2: Streaming inference
// ---------------------------------------------------------------------------

/// Perform streaming inference, printing tokens as they arrive.
async fn example_streaming_inference(client: &AinosClient) -> Result<()> {
    println!("\n=== Streaming Inference ===");

    let req = InferenceRequest::builder()
        .prompt("Tell me a short story about a robot.")
        .model("default")
        .temperature(0.8)
        .max_tokens(500)
        .build();

    let mut stream = client.infer_stream(&req).await?;

    println!("Streaming response:");
    use futures::StreamExt;
    while let Some(chunk) = stream.next().await {
        match chunk {
            Ok(chunk) => {
                print!("{}", chunk.chunk);
                if chunk.done {
                    println!("\n[Stream complete]");
                    break;
                }
            }
            Err(e) => {
                eprintln!("\nStream error: {}", e);
                return Err(e);
            }
        }
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 3: Streaming with collect_string
// ---------------------------------------------------------------------------

/// Use the convenience method to collect a stream into a single string.
async fn example_stream_collect(client: &AinosClient) -> Result<()> {
    println!("\n=== Stream Collect ===");

    let req = InferenceRequest::builder()
        .prompt("Say hello in 3 different languages.")
        .model("default")
        .max_tokens(200)
        .build();

    let mut stream = client.infer_stream(&req).await?;
    let text = stream.collect_string().await?;

    println!("Collected text: {}", text);
    Ok(())
}

// ---------------------------------------------------------------------------
// Example 4: Model management
// ---------------------------------------------------------------------------

/// List loaded models, then load and unload a model.
async fn example_model_management(client: &AinosClient) -> Result<()> {
    println!("\n=== Model Management ===");

    // List models
    println!("Listing models...");
    match client.model_list().await {
        Ok(models) => {
            if models.is_empty() {
                println!("  No models registered.");
            } else {
                for model in &models {
                    println!(
                        "  - {} ({}): {} MB, loaded: {}",
                        model.id, model.name, model.size_mb, model.loaded
                    );
                }
            }
        }
        Err(e) => {
            println!("  Failed to list models: {}", e);
        }
    }

    // Load a model (if path is provided via env)
    if let Ok(model_path) = std::env::var("AINOS_MODEL_PATH") {
        println!("Loading model from: {}", model_path);

        let opts = ModelLoadOptions::builder()
            .path(&model_path)
            .model_type("ggml")
            .gpu_layers(32)
            .context_size(4096)
            .use_mmap(true)
            .build();

        match client.model_load(&model_path, &opts).await {
            Ok(info) => {
                println!("  Model loaded: {} ({})", info.id, info.name);
            }
            Err(e) => {
                println!("  Failed to load model: {}", e);
            }
        }
    } else {
        println!("  Set AINOS_MODEL_PATH to test model loading.");
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 5: Context operations
// ---------------------------------------------------------------------------

/// Store and retrieve context entries.
async fn example_context_operations(client: &AinosClient) -> Result<()> {
    println!("\n=== Context Operations ===");

    let session_id = "example-session";
    let key = "my-key";
    let value = b"Hello from Rust SDK!";

    // Store
    println!("Storing context: {}:{} = {:?}", session_id, key, value);
    client
        .context_store(session_id, key, value, 3600)
        .await?;
    println!("  Stored successfully.");

    // Retrieve
    println!("Retrieving context...");
    match client.context_retrieve(session_id, key).await? {
        Some(data) => {
            let text = String::from_utf8_lossy(&data);
            println!("  Retrieved: {}", text);
        }
        None => {
            println!("  Key not found.");
        }
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 6: Status and health
// ---------------------------------------------------------------------------

/// Query daemon status and health.
async fn example_status(client: &AinosClient) -> Result<()> {
    println!("\n=== Status & Health ===");

    // System status
    let status = client.status().await?;
    println!("System Status:");
    println!("  Uptime: {} seconds", status.uptime);
    println!("  Models loaded: {}", status.models_loaded);
    println!("  Total requests: {}", status.total_requests);
    println!("  Network available: {}", status.network_available);
    println!("  Active sessions: {}", status.active_sessions);

    if let Some(ref limits) = status.rate_limits {
        println!("  Rate Limits:");
        for limit in limits {
            println!(
                "    {}: {}/{} (reset in {}s)",
                limit.category, limit.remaining, limit.limit, limit.reset_seconds
            );
        }
    }

    // Health check
    let health = client.health().await?;
    println!("Health: {}", if health.healthy { "OK" } else { "DEGRADED" });
    println!("  Message: {}", health.message);
    println!("  Database: {}", health.database);
    println!("  Engine: {}", health.engine);
    println!("  Network: {}", health.network);

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 7: Rate limit status
// ---------------------------------------------------------------------------

/// Query rate limit information.
async fn example_rate_limits(client: &AinosClient) -> Result<()> {
    println!("\n=== Rate Limit Status ===");

    match client.rate_limit_status().await {
        Ok(status) => {
            for limit in &status.limits {
                println!(
                    "  {}: {}/{} remaining (resets in {}s)",
                    limit.category, limit.remaining, limit.limit, limit.reset_seconds
                );
            }
        }
        Err(e) => {
            println!("  Failed to query rate limits: {}", e);
        }
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 8: Batch inference
// ---------------------------------------------------------------------------

/// Send multiple inference requests.
async fn example_batch_inference(client: &AinosClient) -> Result<()> {
    println!("\n=== Batch Inference ===");

    let prompts = vec![
        "What is 2+2?",
        "What is the capital of France?",
        "What color is the sky?",
    ];

    let reqs: Vec<InferenceRequest> = prompts
        .into_iter()
        .map(|p| InferenceRequest::builder().prompt(p).max_tokens(100).build())
        .collect();

    let responses = client.batch_infer(&reqs).await?;

    for (i, resp) in responses.iter().enumerate() {
        println!("  [{}] {}...", i + 1, &resp.output[..resp.output.len().min(80)]);
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 9: Error handling
// ---------------------------------------------------------------------------

/// Demonstrates various error handling patterns.
async fn example_error_handling(client: &AinosClient) -> Result<()> {
    println!("\n=== Error Handling ===");

    // Try to unload a non-existent model
    match client.model_unload("nonexistent_model_xyz").await {
        Ok(()) => println!("  Unloaded successfully (unexpected)"),
        Err(e) => {
            println!("  Expected error: {}", e);
            match e.retry_kind() {
                ainos_sdk::RetryKind::Fatal => {
                    println!("    -> Fatal error, not retrying");
                }
                ainos_sdk::RetryKind::Transient => {
                    println!("    -> Transient error, can retry");
                }
                ainos_sdk::RetryKind::Throttled => {
                    println!("    -> Throttled, retry after backoff");
                }
            }
        }
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Example 10: Connection management
// ---------------------------------------------------------------------------

/// Demonstrates explicit connection lifecycle management.
#[allow(dead_code)]
async fn example_connection_management() -> Result<()> {
    println!("\n=== Connection Management ===");

    let client = create_client().await?;
    println!("Connected to {}", client.config().host);

    // Check connection status
    println!("Is connected: {}", client.is_connected().await);
    println!("Is authenticated: {}", client.is_authenticated().await);

    // Disconnect
    client.disconnect().await?;
    println!("Disconnected");

    // Reconnect
    client.reconnect().await?;
    println!("Reconnected");

    Ok(())
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

#[tokio::main]
async fn main() -> Result<()> {
    println!("Ainos SDK Examples");
    println!("==================");

    // Create client
    let client = create_client().await?;

    // Run examples
    example_basic_inference(&client).await?;
    example_streaming_inference(&client).await?;
    example_stream_collect(&client).await?;
    example_model_management(&client).await?;
    example_context_operations(&client).await?;
    example_status(&client).await?;
    example_rate_limits(&client).await?;
    example_batch_inference(&client).await?;
    example_error_handling(&client).await?;

    // Clean up
    client.disconnect().await?;
    println!("\nAll examples completed.");

    Ok(())
}

// ---------------------------------------------------------------------------
// Error handling for main
// ---------------------------------------------------------------------------

/// Run the connection management example independently.
#[allow(dead_code)]
async fn run_connection_example() {
    match example_connection_management().await {
        Ok(()) => println!("Connection example completed."),
        Err(e) => eprintln!("Connection example failed: {}", e),
    }
}